File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第6章：阈值分割/6.2/threshTwoPeaks.py
# -*- coding: utf-8 -*-
import sys
import numpy as np
import cv2
#计算图像灰度直方图
def calcGrayHist(image):
    #灰度图像矩阵的宽高
    rows,cols = image.shape
    #存储灰度直方图
    grayHist = np.zeros([256],np.uint32)
    for r in xrange(rows):
        for c in xrange(cols):
            grayHist[image[r][c]] +=1
    return grayHist
#返回阈值和二值图
def threshTwoPeaks(image):
    #计算回复直方图
    histogram = calcGrayHist(image)
    #找到灰度直方图的最大峰值对应的灰度值
    maxLoc = np.where(histogram==np.max(histogram))
    firstPeak = maxLoc[0]
    print maxLoc[0]
    #寻找灰度直方图的 " 第二个峰值 " 对应的灰度值
    measureDists = np.zeros([256],np.float32)
    for k in xrange(256):
        measureDists[k] = pow(k-firstPeak,2)*histogram[k]
    maxLoc2 = np.where(measureDists==np.max(measureDists))
    secondPeak = maxLoc2[0]
    #找到两个峰值之间的最小值对应的灰度值，作为阈值
    thresh = 0
    if firstPeak > secondPeak:
        temp = histogram[int(secondPeak):int(firstPeak)]
        minLoc = np.where(temp == np.min(temp))
        thresh = secondPeak + minLoc[0]+1
    else:
        temp = histogram[int(firstPeak):int(secondPeak)]
        minLoc = np.where( temp == np.min(temp))
        thresh = firstPeak + minLoc[0]+1
    #找到阈值后进行阈值处理，得到二值图
    threshImage_out = image.copy()
    threshImage_out[threshImage_out > thresh] = 255
    threshImage_out[threshImage_out <= thresh] = 0
    return (thresh,threshImage_out)
#主函数
if __name__ =="__main__":
    if len(sys.argv) > 1:
        image = cv2.imread(sys.argv[1],cv2.CV_LOAD_IMAGE_GRAYSCALE)
    else:
        print "Usge:python threshTwoPeaks.py imageFile"
    thresh,threshImage_out = threshTwoPeaks(image)
    #输出直方图技术得到的阈值
    print thresh
    #显示原图和阈值化得到的二值图
    cv2.imshow("image",image)
    cv2.imshow("threshTwoPeaks",threshImage_out)
    cv2.imwrite("twoPeaks.jpg",threshImage_out)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第6章：阈值分割/6.2/threshTwoPeaks.cpp
#include<iostream>
using namespace std;
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>
using namespace cv;
//计算图像的灰度直方图
Mat calcGrayHist(const Mat & image)
{
	//存储 256 个灰度级的像素数
	Mat histogram = Mat::zeros(Size(256, 1), CV_32SC1);
	int rows = image.rows;
	int cols = image.cols;
	for (int r = 0; r < rows; r++)
	{
		for (int c = 0; c < cols; c++)
		{
			int index = int(image.at<uchar>(r, c));
			histogram.at<int>(0, index) += 1;
		}
	}
	return histogram;
}
//波峰-波谷法
int threshTwoPeaks(const Mat & image,  Mat & thresh_out)
{
	//计算灰度直方图
	Mat histogram = calcGrayHist(image);
	//找到灰度直方图最大峰值对应的灰度级
	Point firstPeakLoc;
	minMaxLoc(histogram, NULL, NULL, NULL, &firstPeakLoc);
	int firstPeak = firstPeakLoc.x;
	//寻找灰度直方图 第二个峰值对应的灰度级
	Mat measureDists = Mat::zeros(Size(256, 1), CV_32FC1);
	for (int k = 0; k < 256; k++)
	{
		int hist_k = histogram.at<int>(0, k);
		measureDists.at<float>(0, k) = pow(float(k - firstPeak), 2)*hist_k;
	}
	Point secondPeakLoc;
	minMaxLoc(measureDists, NULL, NULL, NULL, &secondPeakLoc);
	int secondPeak = secondPeakLoc.x;
	//找到两个峰值之间的最小值对应的灰度级，作为阈值
	Point threshLoc;
	int  thresh = 0;
	if (firstPeak < secondPeak)
	{
		minMaxLoc(histogram.colRange(firstPeak,secondPeak), NULL, NULL, &threshLoc);
		thresh = firstPeak + threshLoc.x + 1;
	}	
	else
	{
		minMaxLoc(histogram.colRange(secondPeak,firstPeak), NULL, NULL, &threshLoc);
		thresh = secondPeak + threshLoc.x + 1;
	}
	//阈值分割
	threshold(image, thresh_out, thresh, 255, cv::THRESH_BINARY);
	return thresh;
}
int main(int argc, char*argv[])
{
	//输入图像矩阵
	Mat image = imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!image.data)
	{
		cout << "没有输入图片" << endl;
		return -1;
	}
	//波峰 波谷阈值法
	Mat threshImage;
	int thresh = threshTwoPeaks(image, threshImage);
	cout << "阈值为：" << thresh << endl;
	//显示阈值后的二值图
	imshow("二值图", threshImage);
	waitKey(0);
	return 0;
}